src/Jackalope/Version/VersionHandler.php

```diff
@@ -235,15 +237,210 @@ private function createFrozenNode(NodeInterface $versionNode, NodeInterface $nod
+                // but if the property exists on the frozen node it has to be copy or version
+                $node->setProperty($propertyName, $property->getValue());
+            }
+        }
+
+        // handle properties present on the node but not on the frozen node
+        foreach ($node->getProperties() as $propertyName => $property) {
+            if ($frozenNode->hasProperty($propertyName)) {
+                continue;
+            }
+
+            /** @var PropertyDefinition $propertyDefinition */
+            $propertyDefinition = $property->getDefinition();
+            $onParentVersion = $propertyDefinition->getOnParentVersion();
+
+            switch ($onParentVersion) {
+                case OnParentVersionAction::COPY:
+                case OnParentVersionAction::VERSION:
+                case OnParentVersionAction::ABORT:
+                    $property->remove();
+                    break;
+                case OnParentVersionAction::INITIALIZE:
+                    $property->setValue($propertyDefinition->determineDefaultValue());
+            }
+        }
+
+        // TODO handle identifier collisions
+
+        // handle child nodes present on the frozen node
+        foreach ($frozenNode->getNodes() as $frozenChildNode) {
+            /** @var NodeInterface $frozenChildNode */
+            if (!$removeExisting) {
+                // TODO check occurence of node in repository
+                throw new NotImplementedException('Check for $removeExisting not implemented yet');
+            }
+
+            $childNodePath = $node->getPath() . '/' . $frozenChildNode->getName();
+            if ($this->session->nodeExists($childNodePath)) {
+                $this->session->removeItem($childNodePath);
+            }
+
+            if (!$frozenChildNode->isNodeType('nt:versionedChild')) {
+                $this->restoreFromNode($node, $frozenChildNode);
+            } else {
+                throw new NotImplementedException('Restoring from nodes with an OPV of VERSION is not implemented.');
+            }
+
+            // TODO remove any node with the same identifier or child identifiers
+        }
+
+        // handle child nodes present on the node but not the frozen node
+        foreach ($node->getNodes() as $childNode) {
+            /** @var NodeInterface $childNode */
+            if ($frozenNode->hasNode($childNode->getName())) {
+                continue;
+            }
+
+            $onParentVersion = $childNode->getDefinition()->getOnParentVersion();
+
+            switch ($onParentVersion) {
+                case OnParentVersionAction::COPY:
+                case OnParentVersionAction::VERSION:
+                case OnParentVersionAction::ABORT:
+                    $childNode->remove();
+                    break;
+                case OnParentVersionAction::INITIALIZE:
+                    // TODO reinitialize node as defined in its node type
```

Comment:
I think that's the only thing left before we can merge this. This behavior here should be similar to the one for properties. The code autocreating the child nodes is currently in the [NodeProcessor](https://github.com/jackalope/jackalope/blob/5f59291a191a29fab16c124ffe725fbe3976f823/src/Jackalope/NodeType/NodeProcessor.php#L158). Should I also extract that to a `determineDefaultValue` to the `NodeDefinition`?